File: mpc/Module_one.m
%% ================================================================
% Module 1: Prediction and system behavior
% Linear system: x(k+1) = A x(k) + B u(k)
% Students change x0, disturbance magnitude & timing, and u.
% =================================================================

clear; 
clc; 
close all;

%% System matrices (2D state vector)

A = [1  0.1; 0   1];   % dynamic matrix
B = [0; 1];            % input matrix

%% Simulation settings

Nsim = 40;             % number of simulation steps
x0  = [-2; 3];         % initial state (students may change)
u   = 0.1;             % constant control input (students may change)

% Add disturbance at a chosen time step:

disturbance_step = 15;  % time step when disturbance is exerted (students may change)
d = [0; 0];             % disturbance vector (default base run: no disturbance)

%% Simulate open-loop trajectory

x = zeros(2 , Nsim + 1);    % initialize predicted state vector with zeros
x(:,1) = x0;                % insert the actual initial state in the first column

for k = 1 : Nsim
    if k == disturbance_step
        x(:,k) = x(:,k) + d;
    end
    x(:,k+1) = A*x(:,k) + B*u;
end

%% Plot

figure; 
hold on; 
grid on;
plot(0 : Nsim, x(1 , :), 'LineWidth', 2)
plot(0 : Nsim, x(2,:), 'LineWidth', 2)
xlabel('$k$','Interpreter','latex'); 
ylabel('State components')
legend('$x_1$','$x_2$','Interpreter','latex')
title('Module 1: Open-loop state prediction')


%% -------- Closed-loop simulation with simple state-feedback K --------
% Activate the following part of the code only for Question 5 (closed-loop preview)

% % Compute feedback gain K via discrete LQR (requires Control System Toolbox)
% % LQR weights for error/state damping (tune if you prefer)
% Q_lqr = diag([1.0, 0.2]);   % penalize position more than velocity
% R_lqr = 0.5;                % penalize input effort moderately
% % Compute stabilizing K (for discrete-time A,B)
% K = dlqr(A, B, Q_lqr, R_lqr);           % K is of size 1 x 2
% 
% % Closed-loop reference (optional): regulate to the origin by default.
% x_ref = [0; 0];   
% 
% x_cl = zeros(2, Nsim+1);
% x_cl(:,1) = x0;
% 
% for k = 1:Nsim
%     % Apply the same disturbance timing/magnitude to compare fairly
%     if k == disturbance_step
%         x_cl(:,k) = x_cl(:,k) + d;
%     end
% 
%     % State-feedback control (regulation to x_ref)
%     u_fb = -K * (x_cl(:,k) - x_ref);   % stabilizing feedback
% 
%     % Closed-loop update
%     x_cl(:,k+1) = A * x_cl(:,k) + B * u_fb;
% end
% 
% % -------------------- Plot: open-loop versus closed loop ---------------------
% kvec = 0:Nsim;
% 
% figure('Color','w'); hold on; grid on;
% 
% % Open-loop
% plot(kvec, x(1,:), 'b-',  'LineWidth', 2, 'DisplayName', '$x_1$ (open-loop)');
% plot(kvec, x(2,:), 'r-',  'LineWidth', 2, 'DisplayName', '$x_2$ (open-loop)');
% 
% % Closed-loop
% plot(kvec, x_cl(1,:), 'b--', 'LineWidth', 2, 'DisplayName', '$x_1$ (closed-loop)');
% plot(kvec, x_cl(2,:), 'r--', 'LineWidth', 2, 'DisplayName', '$x_2$ (closed-loop)');
% 
% % Disturbance marker
% yl = ylim;
% plot([disturbance_step disturbance_step], yl, 'k:', 'LineWidth', 1.5, ...
%      'DisplayName', 'disturbance time');
% 
% xlabel('$k$','Interpreter','latex'); 
% ylabel('State components','Interpreter','latex');
% legend('Interpreter','latex','Location','best');
% title('Module 1: Open-loop vs. closed-loop with disturbance','Interpreter','latex');

File: mpc/Module_one.py
"""Module 1: Prediction and system behavior (Python version).

Python translation of mpc/Module_one.m, including the optional
closed-loop preview requested in Question 5.
"""

from pathlib import Path

import numpy as np
import matplotlib.pyplot as plt

try:
    from scipy.linalg import solve_discrete_are
except Exception as exc:  # pragma: no cover
    raise ImportError(
        "scipy is required for the optional closed-loop preview. "
        "Install with: pip install scipy"
    ) from exc


def format_vec(vec: np.ndarray) -> str:
    return "_".join(f"{float(v):g}".replace("-", "m").replace(".", "p") for v in vec)


# System matrices (2D state vector)
A = np.array([[1.0, 0.1],
              [0.0, 1.0]])
B = np.array([[0.0],
              [1.0]])

# Simulation settings
Nsim = 40
x0 = np.array([-2.0, 3.0])
u = 0.1

# Disturbance setup (assignment base run uses d = [0, 0])
disturbance_step = 15
d = np.array([0.0, 0.0])

# Set to True for Question 5
run_closed_loop_preview = True

# Simulate open-loop trajectory
x = np.zeros((2, Nsim + 1))
x[:, 0] = x0

for k in range(Nsim):
    if (k + 1) == disturbance_step:  # match MATLAB indexing semantics
        x[:, k] = x[:, k] + d
    x[:, k + 1] = A @ x[:, k] + (B[:, 0] * u)


# Plot open-loop trajectory
k = np.arange(Nsim + 1)
fig, ax = plt.subplots(figsize=(8, 4.5))
ax.plot(k, x[0, :], linewidth=2, label="x1")
ax.plot(k, x[1, :], linewidth=2, label="x2")
ax.grid(True)
ax.set_xlabel("k")
ax.set_ylabel("State components")
ax.set_title("Module 1: Open-loop state prediction")
ax.legend()
fig.tight_layout()

fig_dir = Path(__file__).resolve().parent / "figures"
fig_dir.mkdir(parents=True, exist_ok=True)
base_name = (
    f"module1_open_loop_x0_{format_vec(x0)}"
    f"_u_{u:g}_dstep_{disturbance_step}_d_{format_vec(d)}.png"
)
fig_path = fig_dir / base_name
fig.savefig(fig_path, dpi=200, bbox_inches="tight")
print(f"Saved figure: {fig_path}")
if run_closed_loop_preview:
    q_lqr = np.diag([1.0, 0.2])
    r_lqr = np.array([[0.5]])
    p_lqr = solve_discrete_are(A, B, q_lqr, r_lqr)
    k_fb = np.linalg.solve(r_lqr + B.T @ p_lqr @ B, B.T @ p_lqr @ A)

    x_ref = np.array([0.0, 0.0])
    x_cl = np.zeros((2, Nsim + 1))
    x_cl[:, 0] = x0

    for k_idx in range(Nsim):
        if (k_idx + 1) == disturbance_step:
            x_cl[:, k_idx] = x_cl[:, k_idx] + d

        u_fb = -k_fb @ (x_cl[:, k_idx] - x_ref)
        x_cl[:, k_idx + 1] = A @ x_cl[:, k_idx] + B[:, 0] * float(u_fb)

    fig2, ax2 = plt.subplots(figsize=(8.5, 5))
    ax2.plot(k, x[0, :], "b-", linewidth=2, label="x1 (open-loop)")
    ax2.plot(k, x[1, :], "r-", linewidth=2, label="x2 (open-loop)")
    ax2.plot(k, x_cl[0, :], "b--", linewidth=2, label="x1 (closed-loop)")
    ax2.plot(k, x_cl[1, :], "r--", linewidth=2, label="x2 (closed-loop)")
    ax2.axvline(disturbance_step, color="k", linestyle=":", linewidth=1.5, label="disturbance time")
    ax2.grid(True)
    ax2.set_xlabel("k")
    ax2.set_ylabel("State components")
    ax2.set_title("Module 1: Open-loop vs closed-loop with disturbance")
    ax2.legend(loc="best")
    fig2.tight_layout()

    fig_path2 = fig_dir / (
        f"module1_open_vs_closed_x0_{format_vec(x0)}"
        f"_u_{u:g}_dstep_{disturbance_step}_d_{format_vec(d)}.png"
    )
    fig2.savefig(fig_path2, dpi=200, bbox_inches="tight")
    print(f"Saved figure: {fig_path2}")


plt.show()
